src/game_logic/ability_casting.rs

```diff
@@ -105,13 +112,144 @@ impl<'w, 's> AbilityCastingInterface<'w, 's> {
     }
 }
 
+/// Spawns a CastRequest entity for each UseAbility command
+fn request_ability_cast(mut commands: Commands, mut game_commands: MessageReader<GameCommand>) {
+    for command in game_commands.read() {
+        #[expect(irrefutable_let_patterns, reason = "We only have UseAbility currently")]
+        if let GameCommandKind::UseAbility(use_ability) = &command.kind {
+            commands.spawn(use_ability.clone());
+        }
+    }
+}
+
+/// Checks if the ability is on cooldown
+fn check_ability_cooldowns(
+    cast_requests: Query<(Entity, &UseAbility), Without<CastFailed<AbilityCooldown>>>,
+    has_cooldown: Query<Has<Cooldown>>,
+    mut commands: Commands,
+) {
+    for (req_e, use_ability) in cast_requests.iter() {
+        if has_cooldown.get(use_ability.ability_e).unwrap_or(false) {
+            commands
+                .entity(req_e)
+                .insert(CastFailed::<AbilityCooldown>::default());
+        }
+    }
+}
+
+#[derive(Debug, Clone, Reflect)]
+pub struct SlotCooldown;
+
+#[derive(Debug, Clone, Reflect)]
+pub struct SlotRequirement;
+
+/// Checks if the slot is on cooldown
+fn check_slot_cooldowns_real(
+    cast_requests: Query<(Entity, &UseAbility), Without<CastFailed<SlotCooldown>>>,
+    has_cooldown: Query<Has<Cooldown>>,
+    mut commands: Commands,
+) {
+    for (req_e, use_ability) in cast_requests.iter() {
+        if has_cooldown.get(use_ability.slot_e).unwrap_or(false) {
+            commands
+                .entity(req_e)
+                .insert(CastFailed::<SlotCooldown>::default());
+        }
+    }
+}
+
+fn check_slot_requirements(
+    cast_requests: Query<(Entity, &UseAbility), Without<CastFailed<SlotRequirement>>>,
+    abilities: Query<&AbilitySlotRequirement>,
+    slots: Query<&AbilitySlot>,
+    mut commands: Commands,
+) {
+    for (req_e, use_ability) in cast_requests.iter() {
+        let Ok(requirement) = abilities.get(use_ability.ability_e) else {
+            // TODO: Can we enforce this at compile time somehow?
+            error!(
+                "Ability {} has no slot requirement but uses a slot. UseAbility: {use_ability:?}",
+                use_ability.ability_e
+            );
+            continue;
+        };
+
+        let Ok(slot) = slots.get(use_ability.slot_e) else {
+            // Invalid slot entity
+            commands
+                .entity(req_e)
+                .insert(CastFailed::<SlotRequirement>::default());
```

Comment:
This should probably be an `error!()` or something, because it shouldn't happen.